File: C/Linked List/Doubly Circular Linked LIst/DoublyCL02.c
  /*
=====================================================================

Name            : Doubly Circular Linked List all Functions
Program Number  : 002

Description     : This program demonstrates the Count()
                  function of a Doubly Circular Linked List.

                  The function traverses the entire
                  linked list and returns the total
                  number of nodes present.

Concepts Used   : Doubly Circular Linked List,
                  Traversal,
                  Node Counting,
                  do-while Loop

Input           : N/A

Output          : Total number of nodes

Date            : July 2026
Language        : C
Author          : Atharv Tushar Bhosale

=====================================================================
*/

// Doubly Circular Linked list

#include<stdio.h>
#include<stdlib.h>

#pragma pack(1)
struct node
{
    int data;
    struct node *next;
    struct node *prev;
};

typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

void Display(PNODE first, PNODE last)
{

}

int Count(PNODE first, PNODE last)
{
    return 0;
}

void InsertFirst(PPNODE first, PPNODE last, int iNo)
{

}

void InsertLast(PPNODE first, PPNODE last, int iNo)
{

}

void InsertAtPos(PPNODE first, PPNODE last, int iNo, int iPos)
{

}

void DeleteFirst(PPNODE first, PPNODE last)
{

}

void DeleteLast(PPNODE first, PPNODE last)
{

}

void DeleteAtPos(PPNODE first, PPNODE last, int iPos)
{

}

int main()
{
    PNODE head = NULL;
    PNODE tail = NULL;

    return 0;
}

/*
=====================================================================

Sample Output :

Number of nodes are : 0

=====================================================================

Concept Explanation :

Count() traverses the entire linked list
and increments a counter for every node.

Traversal Condition :

while(first != last->next)

This ensures that every node is counted
exactly once in the circular linked list.

=====================================================================
End of Program
=====================================================================
*/

File: Java/Linked List/Doubly Circular Linked List/DoublyCL02.java
/*
===============================================================================
Program Name    : Doubly Circular Linked List - InsertFirst
Program Number  : 002
File Name       : DoublyCL02.java

Description     : This program demonstrates the InsertFirst()
                  operation in a Doubly Circular Linked List 
                  using Java.

Date            : July 2026
Language        : Java
Author          : Atharv Tushar Bhosale
===============================================================================
*/

class node
{
    public int data;
    public node next;
    public node prev;

    node(int no)
    {
        this.data = no;
        this.next = null;
        this.prev = null;
    }
}

class DoublyCL
{
    private node first;
    private node last;
    private int iCount;

    public DoublyCL()
    {
        first = null;
        last = null;
        iCount = 0;
    }

    public void InsertFirst(int iNo)
    {
        node newn = new node(iNo);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            newn.next = first;
            first.prev = newn;
            first = newn;
        }

        last.next = first;
        first.prev = last;

        iCount++;
    }
}

class DoublyCL02
{
    public static void main(String A[])
    {
        DoublyCL dobj = new DoublyCL();

        dobj.InsertFirst(51);
        dobj.InsertFirst(21);
        dobj.InsertFirst(11);

    }
}

/*
===============================================================================
Conclusion:
===============================================================================

Program 002 demonstrates insertion of nodes at the beginning
of a Doubly Circular Linked List using Java.

===============================================================================
*/ 
